# Exploratory Data Analysis of the TMDB Dataset

### Introduction
In this movie dataset analysis, we explore a collection of films spanning various genres, production companies, and release years. The dataset includes key attributes such as budget, revenue, ratings, and the number of reviews, allowing us to investigate trends in box office performance and audience reception. By examining factors like genre popularity, budget-to-revenue ratios, and the impact of critical acclaim, we aim to uncover insights into what drives a movie's commercial and critical success. This analysis provides a comprehensive overview of the film industry's landscape and the characteristics of top-grossing films.

In [1]:
%load_ext sql

In [2]:
%sql sqlite:///TMDB.db

### List All Tables in the Database

In [39]:
%%sql

SELECT name 
FROM sqlite_master 
WHERE type='table';

 * sqlite:///TMDB.db
Done.


[('actors',),
 ('casts',),
 ('genremap',),
 ('genres',),
 ('keywordmap',),
 ('keywords',),
 ('languagemap',),
 ('languages',),
 ('movies',),
 ('oscars',),
 ('productioncompanies',),
 ('productioncompanymap',),
 ('productioncountries',),
 ('productioncountrymap',),
 ('sysdiagrams',)]

## Analyzing the Movies table

In [35]:
%%sql

PRAGMA table_info(movies);

 * sqlite:///TMDB.db
Done.


[(0, 'movie_id', 'INTEGER', 1, None, 1),
 (1, 'title', 'varchar(500)', 0, 'NULL', 0),
 (2, 'release_date', 'datetime(6)', 0, 'NULL', 0),
 (3, 'budget', 'INTEGER', 0, 'NULL', 0),
 (4, 'homepage', 'varchar(500)', 0, 'NULL', 0),
 (5, 'original_language', 'varchar(50)', 0, 'NULL', 0),
 (6, 'original_title', 'varchar(500)', 0, 'NULL', 0),
 (7, 'overview', 'varchar(5000)', 0, 'NULL', 0),
 (8, 'popularity', 'double', 0, 'NULL', 0),
 (9, 'revenue', 'double', 0, 'NULL', 0),
 (10, 'runtime', 'double', 0, 'NULL', 0),
 (11, 'release_status', 'varchar(50)', 0, 'NULL', 0),
 (12, 'tagline', 'varchar(500)', 0, 'NULL', 0),
 (13, 'vote_average', 'double', 0, 'NULL', 0),
 (14, 'vote_count', 'INTEGER', 0, 'NULL', 0),
 (15, 'year_released', 'INT', 0, None, 0)]

#### Find Duplicates

In [207]:
%%sql

SELECT *
FROM movies
GROUP BY title
HAVING COUNT(*) > 1;

 * sqlite:///TMDB.db
Done.


[(268, 'Batman', '1989-06-23 00:00:00.000000', 35000000, None, 'en', 'Batman', "The Dark Knight of Gotham City begins his war on crime with his first major enemy being the clownishly homicidal Joker, who has seized control of Gotham's underworld.", 44.104469, 411348924.0, 126.0, 'Released', 'Have you ever danced with the devil in the pale moonlight?', 7.0, 2096, 1989),
 (10844, 'Out of the Blue', '2006-10-12 00:00:00.000000', None, None, 'en', 'Out of the Blue', 'Ordinary people find extraordinary courage in the face of madness. On 13û14 November 1990 that madness came to Aramoana, a small New Zealand seaside  ... (167 characters truncated) ... hile a handful of under-resourced and under-armed local policemen risked their lives trying to find him and save the survivors. Based on true events.', 0.706355, None, 103.0, 'Released', 'The true story of a small town massacre', 5.9, 18, 2006),
 (1255, 'The Host', '2006-07-27 00:00:00.000000', 11000000, 'http://www.hostmovie.com/', 'ko', '??', "Gang-du is a dim-witted man working at his father's tiny snack bar near the Han River. One day, Gang-du's one and only daughter Hyun-seo comes back f ... (645 characters truncated) ... l of help from Hyun-seo, the once-ordinary citizen Gang-du and his family are thrust into a battle with the monster to rescue their beloved Hyun-seo.", 27.65527, 88489643.0, 119.0, 'Released', 'Monsters are real.', 6.7, 537, 2006)]

3 movies had duplicates in the dataset, however, they were not deleted because further analysis indicated that the movies were released in different years suggesting that the movies had different parts and were not necessarily duplicates.

#### Added a year_released column to the Dataset which is an extraction of the year from the release_date column in the movie table

In [46]:
%%sql

ALTER TABLE movies
ADD COLUMN year_released INT;

 * sqlite:///TMDB.db
(sqlite3.OperationalError) duplicate column name: year_released
[SQL: ALTER TABLE movies
ADD COLUMN year_released INT;]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [37]:
%%sql

UPDATE 
  movies
SET 
  year_released = strftime('%Y', release_date);

 * sqlite:///TMDB.db
4803 rows affected.


[]

#### The total number of records in the movies table

In [208]:
%%sql

SELECT COUNT(*) as no_of_movies
FROM movies;

 * sqlite:///TMDB.db
Done.


[(4801,)]

#### Calculated the minimum, maximum, and average values of the year_released, budget, revenue, runtime, popularity, vote_average and vote_count columns

In [209]:
%%sql
-- Minimum and Maximum values in the year_released column

SELECT 
    MIN(year_released) AS min_year, 
    MAX(year_released) AS max_year
FROM movies;

 * sqlite:///TMDB.db
Done.


[(1916, 2016)]

In [210]:
%%sql
-- Minimum and Maximum values in the budget column

SELECT
    MIN(budget) AS min_budget, 
    MAX(budget) AS max_budget,
    ROUND(AVG(budget)) AS average_budget
FROM movies;

 * sqlite:///TMDB.db
Done.


[(1, 380000000, 37042838.0)]

In [211]:
%%sql
-- Minimum and Maximum values in the revenue column

SELECT 
    MIN(revenue) AS min_revenue, 
    MAX(revenue) AS max_revenue,
    ROUND(AVG(revenue)) AS average_revenue
FROM movies;

 * sqlite:///TMDB.db
Done.


[(5.0, 2787965087.0, 117031353.0)]

In [212]:
%%sql
-- Minimum and Maximum values in the runtime column

SELECT 
    MIN(runtime) AS min_runtime, 
    MAX(runtime) AS max_runtime,
    ROUND(AVG(runtime)) AS average_runtime
FROM movies;

 * sqlite:///TMDB.db
Done.


[(14.0, 338.0, 108.0)]

In [213]:
%%sql
-- Minimum and Maximum values in the popularity column

SELECT 
    MIN(popularity) AS min_popularity, 
    MAX(popularity) AS max_popularity,
    ROUND(AVG(popularity)) AS average_popularity
FROM movies;

 * sqlite:///TMDB.db
Done.


[(0.000372, 875.581305, 22.0)]

In [214]:
%%sql
-- Minimum and Maximum values in the vote_count column

SELECT 
    MIN(vote_count) AS min_vote_count, 
    MAX(vote_count) AS max_vote_count,
    ROUND(AVG(vote_count)) AS average_vote_count
FROM movies;

 * sqlite:///TMDB.db
Done.


[(1, 13752, 699.0)]

In [286]:
%%sql
-- Minimum and Maximum values in the vote_average column

SELECT 
    MIN(vote_average) AS min_vote_average, 
    MAX(vote_average) AS max_vote_average,
    ROUND(AVG(vote_average)) AS average_vote_average
FROM movies;

 * sqlite:///TMDB.db
Done.


[(0.5, 10.0, 6.0)]

In [164]:
%%sql

SELECT COUNT(*) as zero_values FROM movies
WHERE budget = 0 or runtime = 0 or revenue = 0 or popularity 

 * sqlite:///TMDB.db
Done.


[(0,)]

In [94]:
%%sql

SELECT COUNT(*) as zero_values FROM movies
WHERE budget = 0 and runtime = 0 and revenue = 0

 * sqlite:///TMDB.db
Done.


[(31,)]

The analysis indicated that there are are 1573 rows of data that have 0 values in at least on of numerical columns such as budget, runtime, and revenue and 31 rows of that that have 0 values for all three columns which suggests that they could be missing values in the dataset.

In [85]:
%%sql

SELECT * FROM movies
WHERE budget = 0 and runtime = 0 and revenue = 0 and popularity = 0

 * sqlite:///TMDB.db
Done.


[]

In [84]:
%%sql

DELETE FROM movies
WHERE budget = 0 and runtime = 0 and revenue = 0 and popularity = 0

 * sqlite:///TMDB.db
1 rows affected.


[]

The analysis indicated that one movie has zero values for all numerical columns which suggested missing values hence that row was deleted since it is not relevant to our data and might skew the analysis.

In [107]:
%%sql

SELECT COUNT(*) as zero_values FROM movies
WHERE budget = 0

 * sqlite:///TMDB.db
Done.


[(1036,)]

In [108]:
%%sql

UPDATE movies
SET budget = NULLIF(budget, 0),
    runtime = NULLIF(runtime, 0),
    revenue = NULLIF(revenue, 0),
    popularity = NULLIF(popularity, 0),
    vote_count = NULLIF(vote_count, 0),
    vote_average = NULLIF(vote_average, 0);

 * sqlite:///TMDB.db
4802 rows affected.


[]

The Analysis indicated that about 22% of the rows in movies table contained at least one 0 value in the numerical columns i.e. budget, popularity, revenue etc which affected the analysis and range of the data, hence, the zero values were replaced with 'NULL' since it is assumed that those values are missing data.

After replacing the 0 values with 'NULL' the range of the data changed significantly. For instance, the minimum value of the revenue column was 0.00 and the average revenue was 82277769.0 before the 0 values were replaced with 'NULL' and the after the change the values were 5.0 and 117031353.0 respectively.



## Analyzing the rest of the tables in the dataset

In [133]:
%%sql

PRAGMA table_info(genres);

 * sqlite:///TMDB.db
Done.


[(0, 'genre_id', 'INTEGER', 1, None, 1),
 (1, 'genre_name', 'varchar(50)', 0, 'NULL', 0)]

In [135]:
%%sql

PRAGMA table_info(genremap);

 * sqlite:///TMDB.db
Done.


[(0, 'movie_id', 'INTEGER', 1, None, 1),
 (1, 'genre_id', 'INTEGER', 1, None, 2)]

In [136]:
%%sql

PRAGMA table_info(actors);

 * sqlite:///TMDB.db
Done.


[(0, 'actor_id', 'INTEGER', 1, None, 1),
 (1, 'actor_name', 'varchar(100)', 0, 'NULL', 0),
 (2, 'gender', 'INTEGER', 0, 'NULL', 0)]

In [224]:
%%sql

PRAGMA table_info(casts);

 * sqlite:///TMDB.db
Done.


[(0, 'movie_id', 'INTEGER', 1, None, 1),
 (1, 'actor_id', 'INTEGER', 1, None, 2),
 (2, 'characters', 'varchar(500)', 1, None, 3)]

In [137]:
%%sql

PRAGMA table_info(oscars);

 * sqlite:///TMDB.db
Done.


[(0, 'year', 'varchar(10)', 0, 'NULL', 0),
 (1, 'award', 'varchar(500)', 0, 'NULL', 0),
 (2, 'winner', 'varchar(10)', 0, 'NULL', 0),
 (3, 'name', 'varchar(500)', 0, 'NULL', 0),
 (4, 'film', 'varchar(500)', 0, 'NULL', 0)]

In [219]:
%%sql

PRAGMA table_info(productioncompanies);

 * sqlite:///TMDB.db
Done.


[(0, 'production_company_id', 'INTEGER', 1, None, 1),
 (1, 'production_company_name', 'varchar(500)', 0, 'NULL', 0)]

In [222]:
%%sql

PRAGMA table_info(productioncompanymap);

 * sqlite:///TMDB.db
Done.


[(0, 'movie_id', 'INTEGER', 1, None, 1),
 (1, 'production_company_id', 'INTEGER', 1, None, 2)]







# Data Analysis

### What are the top 5 highest-grossing movies in the dataset, and what can be inferred about their success?

In [241]:
%%sql

SELECT 
  title, 
  year_released, 
  genres.genre_name,
  budget, 
  revenue, 
  vote_average as ratings
FROM movies
JOIN genremap ON movies.movie_id = genremap.movie_id
JOIN genres ON genremap.genre_id = genres.genre_id
GROUP BY 
  movies.title, 
  movies.year_released, 
  movies.budget, 
  movies.revenue, 
  movies.vote_average
ORDER BY 
  revenue DESC
LIMIT 5;

 * sqlite:///TMDB.db
Done.


[('Avatar', 2009, 'Adventure', 237000000, 2787965087.0, 7.2),
 ('Titanic', 1997, 'Drama', 200000000, 1845034188.0, 7.5),
 ('The Avengers', 2012, 'Adventure', 220000000, 1519557910.0, 7.4),
 ('Jurassic World', 2015, 'Adventure', 150000000, 1513528810.0, 6.5),
 ('Furious 7', 2015, 'Action', 190000000, 1506249360.0, 7.3)]

#### Key Findings:

- Avatar (2009) and Titanic (1997) are the top-grossing movies, with revenues of 2.78 billion and 1.84 billion, respectively.
- The Adventure genre dominates, with 3 out of 5 movies falling under this category.
- Budget and revenue are strongly correlated, but Furious 7 is an exception, achieving high revenue with a relatively lower budget.
- All movies have high vote averages (6.5-7.5), with Titanic standing out as a particularly well-loved film.
- Most top-grossing movies are recent, with 4 out of 5 released in the 2010s.

### Is there a correlation between a movie's budget and its box office revenue? How strong is this correlation?

In [198]:
%%sql

WITH corr_data AS (
  SELECT 
    budget,
    revenue,
    AVG(budget) OVER () AS avg_budget,
    AVG(revenue) OVER () AS avg_revenue,
    COUNT(*) OVER () AS count
  FROM 
    movies
)
SELECT 
  SUM((budget - avg_budget) * (revenue - avg_revenue)) / 
  ((SUM((budget - avg_budget) * (budget - avg_budget)) * SUM((revenue - avg_revenue) * (revenue - avg_revenue))) * 1.0) AS correlation_coefficient
FROM 
  corr_data;

 * sqlite:///TMDB.db
Done.


[(2.4270143958887464e-20,)]

A correlation coefficient of 2.4270143958887464e-20 is an extremely small value, indicating that there is essentially no correlation between the two variables (budget and revenue). This means that the budget and revenue values do not tend to move together in a predictable way, and there is no significant linear relationship between them.


### What is the trend in the number of movies released each year? Are there noticeable peaks or declines over time?

In [216]:
%%sql

SELECT 
  year_released,
  COUNT(*) AS num_movies
FROM 
  movies
GROUP BY 
  year_released
ORDER BY 
  year_released;

 * sqlite:///TMDB.db
Done.


[(1916, 1),
 (1925, 1),
 (1927, 1),
 (1929, 2),
 (1930, 1),
 (1932, 1),
 (1933, 2),
 (1934, 1),
 (1935, 1),
 (1936, 2),
 (1937, 2),
 (1938, 2),
 (1939, 3),
 (1940, 5),
 (1941, 1),
 (1942, 2),
 (1944, 2),
 (1945, 4),
 (1946, 3),
 (1947, 3),
 (1948, 3),
 (1949, 2),
 (1950, 1),
 (1951, 3),
 (1952, 4),
 (1953, 4),
 (1954, 4),
 (1955, 2),
 (1956, 3),
 (1957, 2),
 (1958, 1),
 (1959, 3),
 (1960, 3),
 (1961, 5),
 (1962, 7),
 (1963, 8),
 (1964, 10),
 (1965, 8),
 (1966, 6),
 (1967, 4),
 (1968, 10),
 (1969, 10),
 (1970, 12),
 (1971, 10),
 (1972, 9),
 (1973, 9),
 (1974, 8),
 (1975, 6),
 (1976, 9),
 (1977, 16),
 (1978, 14),
 (1979, 16),
 (1980, 22),
 (1981, 33),
 (1982, 24),
 (1983, 23),
 (1984, 28),
 (1985, 28),
 (1986, 26),
 (1987, 31),
 (1988, 31),
 (1989, 32),
 (1990, 29),
 (1991, 30),
 (1992, 34),
 (1993, 47),
 (1994, 55),
 (1995, 70),
 (1996, 97),
 (1997, 112),
 (1998, 133),
 (1999, 171),
 (2000, 166),
 (2001, 183),
 (2002, 203),
 (2003, 169),
 (2004, 204),
 (2005, 217),
 (2006, 237),
 (2007, 195),
 (2008, 227),
 (2009, 247),
 (2010, 225),
 (2011, 223),
 (2012, 208),
 (2013, 231),
 (2014, 238),
 (2015, 216),
 (2016, 104)]

The overall trend is an increasing trend, with a few noticeable peaks and declines over time. The number of movies released each year has generally increased over the decades, with some fluctuations.

#### Key Findings:

- Two notable peaks: 1990s-2000s and 2005-2009, with the highest number of movies released in 2009 (247).
- Two notable declines: 1940s-1950s (post-war era) and 2016 (104 movies released).
- Early years: Slow start with few movies released each year.
- Consistency: Despite peaks and declines, the number of movies released has generally increased over time.

#### Possible Factors Influencing the Trend:

- Technological advancements: Improvements in film technology, editing software, and digital distribution platforms might have contributed to the increase in movie production.
- Globalization: The growth of the global film industry, international co-productions, and the increasing popularity of cinema worldwide might have led to an increase in movie production.
- Changes in consumer behavior: Shifts in consumer preferences, such as the rise of streaming services and online content, might have influenced the number of movies released each year.
- Industry trends: Trends within the film industry, such as the rise of franchise films, superhero movies, or independent cinema, might have contributed to the fluctuations in movie production

### Which actors or production companies are associated with the most successful movies in terms of box office revenue and ratings?

In [292]:
%%sql

SELECT 
    movies.title, 
    productioncompanies.production_company_name AS production_companies, 
    actors.actor_name, 
    movies.revenue AS total_revenue, 
    vote_average AS average_ratings
FROM 
    movies
JOIN 
    casts ON movies.movie_id = casts.movie_id
JOIN 
    actors ON casts.actor_id = actors.actor_id
JOIN 
    productioncompanymap ON movies.movie_id = productioncompanymap.movie_id
JOIN 
    productioncompanies ON productioncompanymap.production_company_id = productioncompanies.production_company_id
GROUP BY 
    movies.title
ORDER BY 
    total_revenue DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Avatar', 'Ingenious Film Partners', 'Giovanni Ribisi', 2787965087.0, 7.2),
 ('Titanic', 'Paramount Pictures', 'Kate Winslet', 1845034188.0, 7.5),
 ('The Avengers', 'Paramount Pictures', 'Mark Ruffalo', 1519557910.0, 7.4),
 ('Jurassic World', 'Universal Studios', "Vincent D'Onofrio", 1513528810.0, 6.5),
 ('Furious 7', 'Universal Pictures', 'Lucas Black', 1506249360.0, 7.3),
 ('Avengers: Age of Ultron', 'Marvel Studios', 'Mark Ruffalo', 1405403694.0, 7.3),
 ('Frozen', 'Walt Disney Pictures', 'Edie McClurg', 1274219009.0, 7.3),
 ('Iron Man 3', 'Marvel Studios', 'Mark Ruffalo', 1215439994.0, 6.8),
 ('Minions', 'Universal Pictures', 'Allison Janney', 1156730962.0, 6.4),
 ('Captain America: Civil War', 'Studio Babelsberg', 'William Hurt', 1153304495.0, 7.1)]

#### Key Findings:

- Top-Grossing Movies: The data shows the top 10 highest-grossing movies, with Avatar being the highest-grossing movie with a total revenue of approximately $2.78 billion.

- Production Companies: The data reveals that Universal Studios and Marvel Studios are the most represented production companies in the top 10, with two movies each. Paramount Pictures and Walt Disney Pictures also have two movies each.

- Actors: Mark Ruffalo is the most represented actor in the top 10, with three movies: The Avengers, Avengers: Age of Ultron, and Iron Man 3.

- Revenue and Ratings: There is a general trend of higher revenue corresponding to higher average ratings. However, there are some exceptions, such as Jurassic World, which has a relatively low average rating of 6.5 despite being the 4th highest-grossing movie.

- Genre: Based on the titles, it appears that the top 10 is dominated by superhero movies (The Avengers, Avengers: Age of Ultron, Iron Man 3, Captain America: Civil War) and animated movies (Frozen, Minions)

### How do the average ratings of movies change over different decades or years? Are there any significant shifts in movie quality or audience preferences?

In [255]:
%%sql

SELECT 
  year_released,
  ROUND(AVG(vote_average),1) AS average_rating
FROM 
  movies
GROUP BY 
  year_released
ORDER BY 
  year_released;

 * sqlite:///TMDB.db
Done.


[(1916, 7.4),
 (1925, 7.0),
 (1927, 8.0),
 (1929, 6.3),
 (1930, 6.1),
 (1932, 6.2),
 (1933, 5.6),
 (1934, 7.7),
 (1935, 7.4),
 (1936, 7.5),
 (1937, 7.7),
 (1938, 6.0),
 (1939, 7.7),
 (1940, 6.9),
 (1941, 6.8),
 (1942, 7.4),
 (1944, 6.5),
 (1945, 6.7),
 (1946, 7.3),
 (1947, 6.9),
 (1948, 7.0),
 (1949, 6.7),
 (1950, 7.3),
 (1951, 7.2),
 (1952, 7.2),
 (1953, 6.8),
 (1954, 7.3),
 (1955, 7.4),
 (1956, 6.3),
 (1957, 8.0),
 (1958, 7.6),
 (1959, 6.9),
 (1960, 7.9),
 (1961, 7.3),
 (1962, 7.0),
 (1963, 6.8),
 (1964, 7.2),
 (1965, 6.9),
 (1966, 6.7),
 (1967, 6.8),
 (1968, 6.9),
 (1969, 6.6),
 (1970, 6.7),
 (1971, 6.6),
 (1972, 7.0),
 (1973, 7.0),
 (1974, 7.2),
 (1975, 7.3),
 (1976, 7.1),
 (1977, 6.5),
 (1978, 6.6),
 (1979, 6.5),
 (1980, 6.6),
 (1981, 6.4),
 (1982, 6.4),
 (1983, 6.3),
 (1984, 6.5),
 (1985, 6.5),
 (1986, 6.2),
 (1987, 6.3),
 (1988, 6.3),
 (1989, 6.5),
 (1990, 6.5),
 (1991, 6.2),
 (1992, 6.6),
 (1993, 6.5),
 (1994, 6.3),
 (1995, 6.3),
 (1996, 6.1),
 (1997, 6.3),
 (1998, 6.3),
 (1999, 6.2),
 (2000, 6.0),
 (2001, 6.0),
 (2002, 6.1),
 (2003, 6.1),
 (2004, 6.1),
 (2005, 6.1),
 (2006, 6.2),
 (2007, 6.2),
 (2008, 6.1),
 (2009, 6.1),
 (2010, 6.1),
 (2011, 6.1),
 (2012, 6.0),
 (2013, 6.2),
 (2014, 6.0),
 (2015, 5.9),
 (2016, 5.9)]

The average rating of movies has been declining over the years. The average rating was highest in the early 20th century, with several years having an average rating above 7.5. The rating started to drop in the 1960s and significantly declined in the 1980s. From the 1990s to 2010s, the rating remained stable around 6.1-6.2. However, in the 2010s, the rating further declined, with 2015 and 2016 having the lowest ratings of 5.9.




### What is the distribution of movie budgets across different genres, and how does this impact their average box office performance?

In [5]:
%%sql

-- Step 1: Calculate the distribution of movie budgets across different genres
WITH GenreBudget AS (
    SELECT 
        genres.genre_name, 
        movies.budget,
        movies.revenue
    FROM 
        movies
    JOIN 
        genremap ON movies.movie_id = genremap.movie_id
    JOIN 
        genres ON genremap.genre_id = genres.genre_id
)
-- Step 2: Aggregate the data to analyze the average budget and box office performance per genre
SELECT 
    genre_name,
    COUNT(*) AS num_movies,
    ROUND(AVG(budget)) AS avg_budget,
    ROUND(AVG(revenue)) AS avg_revenue,
    ROUND(AVG(revenue) - AVG(budget)) AS avg_profit_margin,
    MIN(budget) AS min_budget,
    MAX(budget) AS max_budget
FROM 
    GenreBudget
GROUP BY 
    genre_name
ORDER BY 
    avg_revenue DESC;


 * sqlite:///TMDB.db
Done.


[('Animation', 234, 74774140.0, 276503497.0, 201729357.0, 30, 260000000),
 ('Adventure', 790, 72876524.0, 244209721.0, 171333197.0, 1, 380000000),
 ('Fantasy', 424, 71107379.0, 233567521.0, 162460142.0, 1, 380000000),
 ('Family', 512, 63461243.0, 218018949.0, 154557706.0, 30, 260000000),
 ('Science Fiction', 535, 58416989.0, 185795526.0, 127378536.0, 1, 280000000),
 ('Action', 1154, 58564933.0, 173361611.0, 114796679.0, 7, 380000000),
 ('Thriller', 1274, 37956659.0, 107664678.0, 69708019.0, 7, 250000000),
 ('Comedy', 1721, 33249208.0, 104566029.0, 71316821.0, 1, 225000000),
 ('Mystery', 348, 36023843.0, 101674339.0, 65650495.0, 15000, 190000000),
 ('War', 144, 38489954.0, 99331524.0, 60841570.0, 4, 175000000),
 ('Romance', 894, 26703467.0, 88811486.0, 62108019.0, 1, 200000000),
 ('Crime', 696, 32577255.0, 85737171.0, 53159916.0, 7, 250000000),
 ('Drama', 2296, 27408028.0, 78396191.0, 50988163.0, 1, 250000000),
 ('History', 197, 35487849.0, 75547612.0, 40059763.0, 134005, 155000000),
 ('Music', 185, 21481535.0, 74085546.0, 52604011.0, 100, 105000000),
 ('Horror', 519, 18584575.0, 65697368.0, 47112793.0, 1, 200000000),
 ('Western', 82, 33643236.0, 62166707.0, 28523470.0, 200000, 255000000),
 ('Documentary', 110, 5211817.0, 23027185.0, 17815368.0, 2, 60000000),
 ('Foreign', 34, 2796876.0, 2066359.0, -730517.0, 7, 7000000),
 ('TV Movie', 8, 3066667.0, None, None, 1000000, 4200000)]

#### Key Findings:
- Highest-grossing genres: The top 3 genres by average revenue are Animation, Adventure, and Fantasy, with average revenues exceeding 200 million.
- Lowest-grossing genres: The bottom 3 genres by average revenue are Foreign, TV Movie, and Documentary, with average revenues below 25 million.
- Highest-budget genres: The top 3 genres by average budget are Animation, Adventure, and Fantasy, with average budgets exceeding 70 million.
- Lowest-budget genres: The bottom 3 genres by average budget are Foreign, TV Movie, and Documentary, with average budgets below 5 million.
- Most profitable genres: The top 3 genres by average profit margin are Animation, Adventure, and Fantasy, with average profit margins exceeding 150 million.
- Least profitable genres: The bottom 3 genres by average profit margin are Foreign, TV Movie, and Documentary, with average profit margins below 10 million or even negative.

#### Genre-specific insights

- Animation: Has the highest average budget and revenue, with a strong profit margin. This suggests that animation movies are often high-investment, high-reward projects.
- Action: Has a relatively high average budget, but a lower average revenue compared to other high-budget genres. This might indicate that action movies are more hit-or-miss in terms of box office performance.
- Comedy: Has a relatively low average budget, but a moderate average revenue. This suggests that comedies can be profitable with lower investments.
- Horror: Has a relatively low average budget and revenue, but a decent profit margin. This might indicate that horror movies can be profitable with lower investments and a focus on niche audiences.
- Documentary: Has a very low average budget and revenue, with a negative profit margin. This suggests that documentaries are often low-budget, low-revenue projects that may not be as commercially viable.

#### Correlations and trends

- Budget vs. Revenue: There is a strong positive correlation between average budget and average revenue across genres. This suggests that higher budgets often lead to higher revenues.
- Budget vs. Profit Margin: There is a moderate positive correlation between average budget and average profit margin across genres. This suggests that higher budgets can lead to higher profit margins, but not always.
- Genre diversity: The data shows a diverse range of genres, with varying levels of budget, revenue, and profit margin. This suggests that different genres have different commercial viability and profitability profiles.

### Do movies with a higher number of reviews tend to have higher ratings? What is the relationship between the number of reviews and the average rating?

In [265]:
%%sql

SELECT title, vote_count AS number_of_reviews, vote_average AS ratings
FROM movies
ORDER BY number_of_reviews DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Inception', 13752, 8.1),
 ('The Dark Knight', 12002, 8.2),
 ('Avatar', 11800, 7.2),
 ('The Avengers', 11776, 7.4),
 ('Deadpool', 10995, 7.4),
 ('Interstellar', 10867, 8.1),
 ('Django Unchained', 10099, 7.8),
 ('Guardians of the Galaxy', 9742, 7.9),
 ('The Hunger Games', 9455, 6.9),
 ('Mad Max: Fury Road', 9427, 7.2)]

In [270]:
%%sql

WITH 
  -- Calculate means
  means AS (
    SELECT 
      AVG(vote_count) AS avg_reviews,
      AVG(vote_average) AS avg_rating
    FROM 
      movies
  ),
  
  -- Calculate deviations and products
  deviations AS (
    SELECT 
      vote_count - m.avg_reviews AS reviews_dev,
      vote_average - m.avg_rating AS rating_dev,
      (vote_count - m.avg_reviews) * (vote_average - m.avg_rating) AS product,
      (vote_count - m.avg_reviews) * (vote_count - m.avg_reviews) AS reviews_dev_squared,
      (vote_average - m.avg_rating) * (vote_average - m.avg_rating) AS rating_dev_squared
    FROM 
      movies
    CROSS JOIN means m
  )

SELECT 
  SUM(product) AS sum_product,
  SUM(reviews_dev_squared) AS sum_reviews_dev_squared,
  SUM(rating_dev_squared) AS sum_rating_dev_squared
FROM 
  deviations;

 * sqlite:///TMDB.db
Done.


[(1949145.9901033984, 7288817125.95358, 4482.125667862438)]

In [271]:
import sqlite3
import math

# Connect to the database
conn = sqlite3.connect('TMDB.db')
cursor = conn.cursor()

# Execute the SQL query
query = """
WITH 
  means AS (
    SELECT 
      AVG(vote_count) AS avg_reviews,
      AVG(vote_average) AS avg_rating
    FROM 
      movies
  ),
  deviations AS (
    SELECT 
      vote_count - m.avg_reviews AS reviews_dev,
      vote_average - m.avg_rating AS rating_dev,
      (vote_count - m.avg_reviews) * (vote_average - m.avg_rating) AS product,
      (vote_count - m.avg_reviews) * (vote_count - m.avg_reviews) AS reviews_dev_squared,
      (vote_average - m.avg_rating) * (vote_average - m.avg_rating) AS rating_dev_squared
    FROM 
      movies
    CROSS JOIN means m
  )
SELECT 
  SUM(product) AS sum_product,
  SUM(reviews_dev_squared) AS sum_reviews_dev_squared,
  SUM(rating_dev_squared) AS sum_rating_dev_squared
FROM 
  deviations;
"""

cursor.execute(query)
result = cursor.fetchone()

# Retrieve the sums from the result
sum_product = result[0]
sum_reviews_dev_squared = result[1]
sum_rating_dev_squared = result[2]

# Calculate the correlation coefficient
correlation_coefficient = sum_product / math.sqrt(sum_reviews_dev_squared * sum_rating_dev_squared)

print("Correlation Coefficient:", correlation_coefficient)

# Close the connection
conn.close()

Correlation Coefficient: 0.3410152707876165


#### Insights:

The correlation coefficient of 0.34 suggests a moderate positive relationship between the number of reviews and average ratings, implying that movies with more reviews tend to have slightly higher ratings, though the relationship is not particularly strong.
This indicaties that there are many other factors influencing the rating beyond just the number of reviews. This is expected, as movie ratings are influenced by a complex array of factors, such as the movie's genre, director, cast, plot, and overall quality.

### What is the average rating of movies in the dataset?

In [278]:
%%sql

SELECT ROUND(AVG(vote_average),1) as average_rating_of_movies
FROM movies;

 * sqlite:///TMDB.db
Done.


[(6.2,)]

### Which genre of movies has the highest average rating?

In [283]:
%%sql

SELECT 
  genres.genre_name,
  AVG(movies.vote_average) as average_ratings
FROM movies
JOIN genremap ON movies.movie_id = genremap.movie_id
JOIN genres ON genremap.genre_id = genres.genre_id
GROUP BY genres.genre_name
ORDER BY average_ratings DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('History', 6.71979695431472),
 ('War', 6.713888888888885),
 ('Documentary', 6.662135922330096),
 ('Drama', 6.449956024626197),
 ('Music', 6.425136612021856),
 ('Western', 6.412658227848103),
 ('Foreign', 6.35294117647059),
 ('Animation', 6.341452991452991),
 ('Crime', 6.319536903039073),
 ('Mystery', 6.237681159420291)]

#### Insights:

- Historical dramas and war movies tend to be well-received, possibly due to their often-serious and thought-provoking nature.
- Documentaries, which often focus on real-life events or issues, may appeal to audiences seeking informative and engaging content.
- Drama and Music, may be more hit-or-miss, with some movies resonating with audiences and others not as much.
- Animation, Crime, and Mystery movies may struggle to connect with audiences, potentially due to oversaturation or a lack of originality in these genres.

### What is the relationship between popularity and revenue?

In [12]:
%%sql

SELECT title, revenue, popularity
FROM movies
ORDER BY revenue DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Avatar', 2787965087.0, 150.437577),
 ('Titanic', 1845034188.0, 100.025899),
 ('The Avengers', 1519557910.0, 144.448633),
 ('Jurassic World', 1513528810.0, 418.708552),
 ('Furious 7', 1506249360.0, 102.322217),
 ('Avengers: Age of Ultron', 1405403694.0, 134.279229),
 ('Frozen', 1274219009.0, 165.125366),
 ('Iron Man 3', 1215439994.0, 77.68208),
 ('Minions', 1156730962.0, 875.581305),
 ('Captain America: Civil War', 1153304495.0, 198.372395)]

In [16]:
%%sql

WITH 
  -- Calculate the mean of popularity and revenue
  means AS (
    SELECT 
      AVG(popularity) AS mean_popularity,
      AVG(revenue) AS mean_revenue
    FROM 
      movies
  ),
  
  -- Calculate the deviations from the mean
  deviations AS (
    SELECT 
      popularity - mean_popularity AS popularity_dev,
      revenue - mean_revenue AS revenue_dev
    FROM 
      movies
    CROSS JOIN means
  ),
  
  -- Calculate the covariance
  covariance AS (
    SELECT 
      SUM(popularity_dev * revenue_dev) AS covariance
    FROM 
      deviations
  ),
    sum_squares AS (
    SELECT 
      SUM(popularity_dev * popularity_dev) AS sum_popularity_squares,
      SUM(revenue_dev * revenue_dev) AS sum_revenue_squares
    FROM 
      deviations
  )

-- Calculate the correlation coefficient
SELECT 
  covariance / (sum_popularity_squares * sum_revenue_squares) AS correlation_coefficient
FROM 
  covariance
  CROSS JOIN sum_squares;

 * sqlite:///TMDB.db
Done.


[(2.4207179099259195e-14,)]

The correlation coefficient of 2.4207179099259195e-14 is extremely close to zero, which indicates that there is almost no linear relationship between the popularity and revenue variables.

### What is the average revenue of movies in the dataset?

In [285]:
%%sql

SELECT ROUND(AVG(revenue),1) as average_revenue
FROM movies;

 * sqlite:///TMDB.db
Done.


[(117031352.9,)]

### Which production company has made the most movies in the dataset?

In [6]:
%%sql

SELECT productioncompanies.production_company_name AS production_companies, COUNT(movies.title) as number_of_movies, SUM(revenue) as revenue
FROM movies
JOIN productioncompanymap ON movies.movie_id = productioncompanymap.movie_id
JOIN productioncompanies ON productioncompanymap.production_company_id = productioncompanies.production_company_id
GROUP BY productioncompanies.production_company_name
ORDER BY revenue DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Warner Bros.', 319, 49155747874.0),
 ('Universal Pictures', 311, 42588465942.0),
 ('Paramount Pictures', 285, 40878523165.0),
 ('Twentieth Century Fox Film Corporation', 222, 39357151309.0),
 ('Walt Disney Pictures', 114, 28683256048.0),
 ('Columbia Pictures', 201, 28599634775.0),
 ('New Line Cinema', 165, 19444865804.0),
 ('Amblin Entertainment', 49, 16090835147.0),
 ('DreamWorks SKG', 79, 14384533626.0),
 ('Dune Entertainment', 59, 13797504190.0)]

#### Insights:

The data indicates that Warner Bros. and Universal Pictures have produced the most movies in the dataset, while Relativity Media is the least active of the top 10 production companies. 

### Which actor has appeared in the most movies in the dataset?

In [294]:
%%sql
SELECT 
    actors.actor_name, 
    COUNT(movies.title) as number_of_movies
FROM movies
JOIN casts ON movies.movie_id = casts.movie_id
JOIN actors ON casts.actor_id = actors.actor_id
GROUP BY actors.actor_name
ORDER BY  number_of_movies DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Samuel L. Jackson', 67),
 ('Robert De Niro', 57),
 ('Bruce Willis', 51),
 ('Matt Damon', 48),
 ('Morgan Freeman', 46),
 ('Steve Buscemi', 43),
 ('Liam Neeson', 41),
 ('Owen Wilson', 40),
 ('Johnny Depp', 40),
 ('Nicolas Cage', 39)]

### Findings:

- Samuel L. Jackson is the clear leader, having appeared in 67 movies, a significant margin ahead of the next closest actor, Robert De Niro, who has appeared in 57 movies.
- The top 3 actors (Jackson, De Niro, and Bruce Willis) have all appeared in over 50 movies, indicating a high level of prolificacy and longevity in their careers.
- The middle tier of actors (Matt Damon to Owen Wilson) have all appeared in between 40-48 movies, suggesting a consistent level of activity in the film industry.
- Johnny Depp and Nicolas Cage, both known for their versatility and range, round out the top 10 with 40 and 39 movie appearances, respectively.

### What is the relationship between movie budget and rating? Do higher-budget movies tend to have higher ratings?

In [298]:
%%sql

SELECT title, budget, vote_average AS ratings
FROM movies
ORDER BY budget DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Pirates of the Caribbean: On Stranger Tides', 380000000, 6.4),
 ("Pirates of the Caribbean: At World's End", 300000000, 6.9),
 ('Avengers: Age of Ultron', 280000000, 7.3),
 ('Superman Returns', 270000000, 5.4),
 ('Tangled', 260000000, 7.4),
 ('John Carter', 260000000, 6.1),
 ('Spider-Man 3', 258000000, 5.9),
 ('The Lone Ranger', 255000000, 5.9),
 ('Harry Potter and the Half-Blood Prince', 250000000, 7.4),
 ('The Dark Knight Rises', 250000000, 7.6)]

In [303]:
%%sql

WITH 
  -- Calculate the average budget and rating
  averages AS (
    SELECT 
      AVG(budget) AS avg_budget,
      AVG(vote_average) AS avg_rating
    FROM 
      movies
  ),
  
  -- Calculate the deviations from the average
  deviations AS (
    SELECT 
      budget - a.avg_budget AS budget_dev,
      vote_average - a.avg_rating AS rating_dev
    FROM 
      movies
    CROSS JOIN averages a
  ),
  
  -- Calculate the covariance and variance
  covariance_variance AS (
    SELECT 
      SUM(budget_dev * rating_dev) AS covariance,
      SUM(budget_dev * budget_dev) AS variance_budget,
      SUM(rating_dev * rating_dev) AS variance_rating
    FROM 
      deviations
  )
  
-- Calculate the correlation coefficient
SELECT 
  covariance / ((variance_budget * variance_rating) * (1.0/2)) AS correlation_coefficient
FROM 
  covariance_variance;

 * sqlite:///TMDB.db
Done.


[(3.641432379766429e-14,)]

#### Insights;

The correlation coefficient of 3.64e-14 is an extremely small value close to zero. This suggests that there is no significant linear relationship between the two variables (budget and rating) in the dataset.

In other words, the correlation coefficient is indicating that the budget and rating variables are essentially unrelated, and changes in one variable do not tend to be associated with changes in the other variable.

### Is there a correlation between movie runtime and rating? Do longer movies tend to have higher ratings?

In [304]:
%%sql

SELECT title, runtime, vote_average AS ratings
FROM movies
ORDER BY runtime DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Carlos', 338.0, 6.7),
 ('The Company', 276.0, 7.0),
 ('Gettysburg', 254.0, 6.6),
 ('Cleopatra', 248.0, 6.7),
 ('Hamlet', 242.0, 7.3),
 ('Emma', 240.0, 7.6),
 ('Gone with the Wind', 238.0, 7.7),
 ('Once Upon a Time in America', 229.0, 8.2),
 ('Woodstock', 225.0, 7.1),
 ("Heaven's Gate", 219.0, 6.4)]

In [307]:
%%sql

WITH 
  -- Calculate the average runtime and rating
  averages AS (
    SELECT 
      AVG(runtime) AS avg_runtime,
      AVG(vote_average) AS avg_rating
    FROM 
      movies
  ),
  
  -- Calculate the deviations from the average
  deviations AS (
    SELECT 
      runtime - a.avg_runtime AS runtime_dev,
      vote_average - a.avg_rating AS rating_dev
    FROM 
      movies
    CROSS JOIN averages a
  ),
  
  -- Calculate the covariance and variance
  covariance_variance AS (
    SELECT 
      SUM(runtime_dev * rating_dev) AS covariance,
      SUM(runtime_dev * runtime_dev) AS variance_runtime,
      SUM(rating_dev * rating_dev) AS variance_rating
    FROM 
      deviations
  )
  
-- Calculate the correlation coefficient
SELECT 
  covariance / ((variance_runtime * variance_rating) * (1.0/2)) AS correlation_coefficient
FROM 
  covariance_variance;

 * sqlite:///TMDB.db
Done.


[(7.394180314011995e-06,)]

#### Insights:

The correlation coefficient of 7.394180314011995e-06 suggests that there is no significant linear relationship between the movie runtime and rating.This means that the movie runtime and rating are not significantly related, and changes in one variable do not tend to be associated with changes in the other variable.

### Which actor is the most awarded in the dataset?

In [8]:
%%sql

SELECT 
    actors.actor_name, 
    COUNT(oscars.winner) as number_of_oscars
FROM movies
JOIN casts ON movies.movie_id = casts.movie_id
JOIN actors ON casts.actor_id = actors.actor_id
JOIN oscars ON actors.actor_name = oscars.name
GROUP BY actors.actor_name
ORDER BY  number_of_oscars DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Robert De Niro', 114),
 ('Meryl Streep', 90),
 ('Tom Hanks', 64),
 ('Denzel Washington', 62),
 ('Cate Blanchett', 62),
 ('Michael Caine', 58),
 ('Dustin Hoffman', 58),
 ('Jack Nicholson', 51),
 ('Gene Hackman', 50),
 ('Angelina Jolie', 50)]

### Which actor has the most oscar nomination?

In [7]:
%%sql

SELECT 
    actors.actor_name, 
    COUNT(oscars.name) as number_of_nominations
FROM movies
JOIN casts ON movies.movie_id = casts.movie_id
JOIN actors ON casts.actor_id = actors.actor_id
JOIN oscars ON actors.actor_name = oscars.name
GROUP BY actors.actor_name
ORDER BY  number_of_nominations DESC
LIMIT 10;

 * sqlite:///TMDB.db
Done.


[('Meryl Streep', 570),
 ('Robert De Niro', 399),
 ('Morgan Freeman', 230),
 ('Cate Blanchett', 217),
 ('Robert Duvall', 210),
 ('Judi Dench', 210),
 ('Jack Nicholson', 204),
 ('Dustin Hoffman', 203),
 ('Denzel Washington', 186),
 ('Al Pacino', 184)]

### Convert the dataset to csv

In [218]:
import sqlite3
import csv

# Connect to the database
conn = sqlite3.connect('TMDB.db')
cursor = conn.cursor()

# Get a list of all tables in the database
cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
tables = [row[0] for row in cursor.fetchall()]

# Loop through each table and export it to a separate CSV file
for table in tables:
    cursor.execute(f"SELECT * FROM {table}")
    rows = cursor.fetchall()
    with open(f"{table}.csv", 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow([desc[0] for desc in cursor.description])
        writer.writerows(rows)

# Close the database connection
conn.close()